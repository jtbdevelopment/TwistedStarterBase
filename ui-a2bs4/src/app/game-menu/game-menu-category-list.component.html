<div *ngIf="games.length > 0" class="category-wrapper">
    <button class="list-group-item btn-group btn btn-group-{{style}}" (click)="toggleCollapse()">
        <span class="fa fa-{{icon}} float-left"></span>
        <span class="float-left">{{category}}</span>
        <span class="badge badge-secondary badge-pill align-middle float-right ml-auto">{{games.length}}</span>
    </button>
    <div [ngbCollapse]="isCollapsed" class="games-collapse">
        <div *ngFor="let game of games;" class="btn-group-vertical btn-group-games">
            <game-menu-game-item [game]="game" [style]="style" class="game-wrapper"></game-menu-game-item>
        </div>
    </div>
</div>
