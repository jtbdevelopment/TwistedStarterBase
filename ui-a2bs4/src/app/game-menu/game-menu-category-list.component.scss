.category-wrapper {
  width: 100%;

  .fa {
    margin-right: 5px;
    margin-top: 4px;
  }

  .list-group-item {
    width: 100%;
    margin-top: 5px;
    -ms-word-wrap: break-word;
    word-wrap: break-word;
    white-space: normal;
    font-weight: bold;
    padding: 10px 10px;
    border: none;
    color: gray;
    .badge {
      margin-top: 4px;
    }
  }

  .games-collapse {
    width: 100%;
    .btn-group-games {
      width: 100%;
      .game-wrapper {
        width: 100%;
      }
    }
  }
}

