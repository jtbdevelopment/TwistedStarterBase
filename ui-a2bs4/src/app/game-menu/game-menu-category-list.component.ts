import {Component, Input, OnInit} from '@angular/core';
import {Game, GameCacheService} from 'jtb-core-games-ui';


@Component({
  selector: 'app-game-menu-category-list',
  templateUrl: './game-menu-category-list.component.html',
  styleUrls: ['./game-menu-category-list.component.scss']
})
export class GameMenuCategoryListComponent implements OnInit {
  @Input() public category: string;
  @Input() public icon: string;
  @Input() public style: string;
  public games: Game[] = [];
  public isCollapsed: boolean = false;

  constructor(private gameCache: GameCacheService) {
  }

  ngOnInit() {
    this.gameCache.getGamesForCategory(this.category).subscribe(g => {
      this.games = Object.assign([], g);
      this.games.sort((a, b) => {
        if (a.lastUpdate > b.lastUpdate) {
          return 1;
        }
        if (a.lastUpdate === b.lastUpdate) {
          return 0;
        }
        return -1;
      });
    });
  }

  public toggleCollapse(): void {
    this.isCollapsed = !this.isCollapsed;
  }
}
