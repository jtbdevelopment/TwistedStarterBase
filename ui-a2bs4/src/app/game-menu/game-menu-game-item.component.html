<a [routerLink]="[game.standardLink()]"
   class="btn-group btn btn-light btn-group-game btn-group-game-{{style}}">
  <img *ngFor="let md5 of game.players | mapKeys" class="player-image" src="{{game.playerImages[md5]}}">
  <span class="game-description">{{describeGame()}}</span>
</a>

