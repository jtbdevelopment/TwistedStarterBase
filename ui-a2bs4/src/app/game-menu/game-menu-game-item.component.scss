@import '~bootstrap/scss/bootstrap.scss';

.btn-group-game {
  width: 100%;
  padding: 4px;
  alignment: left;
  text-align: left;
  -ms-word-wrap: break-word;
  word-wrap: break-word;
  white-space: normal;
  font-weight: bold;
  border: 1px solid;
  .game-description {
    font-weight: bold;
  }

  .player-image {
    margin: 3px 1px;
    width: 25px;
    height: 25px;
    @include media-breakpoint-down(sm) {
      width: 20px;
      height: 20px;
    }
  }
}
