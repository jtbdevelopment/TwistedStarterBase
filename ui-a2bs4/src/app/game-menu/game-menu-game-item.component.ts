import {Component, Input} from '@angular/core';
import {Game} from '../core-ui/games/game.model';
@Component({
    selector: 'game-menu-game-item',
    template: require('./game-menu-game-item.component.html')
})
export class GameMenuGameItemComponent {
    @Input() public game: Game;
    @Input() public style: string;

    //  TODO - TSB
    public describeGame(): string {
        return 'TODO';
    }
}
