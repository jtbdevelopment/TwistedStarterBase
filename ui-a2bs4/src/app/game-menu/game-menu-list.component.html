<div class="game-bar">
    <div *ngFor="let category of categories; let i = index;"
         class="list-group list-group-games list-group-{{styles[i]}}">
        <game-menu-category-list [category]="category"
                                 [icon]="icons.get(category)"
                                 [style]="styles[i]">
        </game-menu-category-list>
    </div>
</div>