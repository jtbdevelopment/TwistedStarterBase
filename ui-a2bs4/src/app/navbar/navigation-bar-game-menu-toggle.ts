import {Component, Input} from '@angular/core';
import {GameMenuService} from '../game-menu/gamemenuservice';

@Component({
    selector: 'navigation-bar-game-menu-toggle',
    template: require('./navigation-bar-game-menu-toggle.html')
})
export class NavigationBarGameMenuToggleComponent {
    @Input() playerLoaded: boolean;

    private beforeHoverMenuValue: boolean;

    constructor(private gameMenuService: GameMenuService) {
    }

    hoverGameMenu(): void {
        this.beforeHoverMenuValue = this.gameMenuService.getShowGames();
        this.gameMenuService.setShowGames(true);
    }

    stopHoverGameMenu(): void {
        this.gameMenuService.setShowGames(this.beforeHoverMenuValue);
    }

    toggleGameMenu(): void {
        this.gameMenuService.setShowGames(!this.gameMenuService.getShowGames());
    }
}
