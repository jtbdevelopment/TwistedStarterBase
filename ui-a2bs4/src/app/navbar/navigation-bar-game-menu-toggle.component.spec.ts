import {inject, TestBed} from '@angular/core/testing';
import {NavigationBarGameMenuToggleComponent} from './navigation-bar-game-menu-toggle.component';
import {GameMenuService} from '../game-menu/game-menu.service';
import {HelpDisplayService} from '../help/help-display.service';
import {NgbModule, NgbPopoverConfig} from '@ng-bootstrap/ng-bootstrap';
import {AppConfig} from 'jtb-core-games-ui';
import {Observable} from 'rxjs/Observable';
import {BehaviorSubject} from 'rxjs/BehaviorSubject';
import {from} from 'rxjs/observable/from';


export class MockGameMenuService {
    static showGamesSubject: BehaviorSubject<boolean> = new BehaviorSubject(true);

    //noinspection JSUnusedGlobalSymbols
    public showGames: Observable<boolean> = from(MockGameMenuService.showGamesSubject);

    //noinspection JSUnusedGlobalSymbols,JSMethodCanBeStatic
    setShowGames(show: boolean): void {
        MockGameMenuService.showGamesSubject.next(show);
    }

    //noinspection JSUnusedGlobalSymbols,JSMethodCanBeStatic
    getShowGames(): boolean {
        return MockGameMenuService.showGamesSubject.getValue();
    }
}

export class MockAppConfig implements AppConfig {
    appName = 'Twisted Starter Base';
    hoverMenu = true;
    version = '';
    releaseNotes = '';
    // noinspection JSUnusedGlobalSymbols
    inviteFriendsMessage = '';
}

describe('Component:  nav bar game menu toggle component', () => {
    beforeEach(() => {
        TestBed.configureTestingModule({
            imports: [
                NgbModule
            ],
            declarations: [
                NavigationBarGameMenuToggleComponent,
            ],
            providers: [
                {provide: GameMenuService, useClass: MockGameMenuService},
                {provide: 'AppConfig', useClass: MockAppConfig},
                HelpDisplayService,
                NgbPopoverConfig
            ],
        });
        TestBed.compileComponents();
    });

    it('displays no toggle when player loaded', () => {
        const fixture = TestBed.createComponent(NavigationBarGameMenuToggleComponent);
        fixture.componentInstance.playerLoaded = false;
        fixture.detectChanges();
        expect(fixture).toMatchSnapshot();
    });

    it('displays toggle when player loaded', () => {
        const fixture = TestBed.createComponent(NavigationBarGameMenuToggleComponent);
        fixture.componentInstance.playerLoaded = true;
        fixture.detectChanges();
        expect(fixture).toMatchSnapshot();
    });

    it('can toggle menu on/off from non-hover', () => {
        const fixture = TestBed.createComponent(NavigationBarGameMenuToggleComponent);
        fixture.detectChanges();
        expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
        fixture.componentInstance.toggleGameMenu();
        expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
        fixture.componentInstance.toggleGameMenu();
        expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
        fixture.componentInstance.toggleGameMenu();
        expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
    });

    it('toggling help on/off shows/hides popover', inject([HelpDisplayService], (helpService) => {
        const fixture = TestBed.createComponent(NavigationBarGameMenuToggleComponent);
        fixture.componentInstance.playerLoaded = true;
        fixture.detectChanges();
        helpService.toggleHelp();
        fixture.detectChanges();
        expect(fixture).toMatchSnapshot();
        helpService.toggleHelp();
        fixture.detectChanges();
        expect(fixture).toMatchSnapshot();
    }));

    describe('when enableHover is true', () => {
        beforeEach(inject(['AppConfig'], (config) => {
            config.hoverMenu = true;
        }));
        it('hovering on forces show games to true', () => {
            const fixture = TestBed.createComponent(NavigationBarGameMenuToggleComponent);
            fixture.detectChanges();
            MockGameMenuService.showGamesSubject.next(false);
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
            fixture.componentInstance.hoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
        });

        it('stopping hovering on restores show games to true if started true', () => {
            const fixture = TestBed.createComponent(NavigationBarGameMenuToggleComponent);
            fixture.detectChanges();
            MockGameMenuService.showGamesSubject.next(true);
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
            fixture.componentInstance.hoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
            fixture.componentInstance.stopHoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
        });

        it('stopping hovering on restores show games to false if started false', () => {
            const fixture = TestBed.createComponent(NavigationBarGameMenuToggleComponent);
            fixture.detectChanges();
            MockGameMenuService.showGamesSubject.next(false);
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
            fixture.componentInstance.hoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
            fixture.componentInstance.stopHoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
        });

        it('toggling while hovering finalizes state - assume starting with hide menu', () => {
            const fixture = TestBed.createComponent(NavigationBarGameMenuToggleComponent);
            fixture.detectChanges();
            MockGameMenuService.showGamesSubject.next(false);
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
            fixture.componentInstance.hoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
            fixture.componentInstance.toggleGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
            fixture.componentInstance.stopHoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
        });

        it('toggling while hovering finalizes state - assume starting with showing menu', () => {
            const fixture = TestBed.createComponent(NavigationBarGameMenuToggleComponent);
            fixture.detectChanges();
            MockGameMenuService.showGamesSubject.next(true);
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
            fixture.componentInstance.hoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
            fixture.componentInstance.toggleGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
            fixture.componentInstance.stopHoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
        });
    });

    describe('when enableHover is false', () => {
        beforeEach(inject(['AppConfig'], (config) => {
            config.hoverMenu = false;
        }));
        it('hovering on forces does not enable show games', () => {
            const fixture = TestBed.createComponent(NavigationBarGameMenuToggleComponent);
            fixture.detectChanges();
            MockGameMenuService.showGamesSubject.next(false);
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
            fixture.componentInstance.hoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
        });

        it('stopping hovering on restores show games to true if started true', () => {
            const fixture = TestBed.createComponent(NavigationBarGameMenuToggleComponent);
            fixture.detectChanges();
            MockGameMenuService.showGamesSubject.next(true);
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
            fixture.componentInstance.hoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
            fixture.componentInstance.stopHoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
        });

        it('stopping hovering on restores show games to false if started false', () => {
            const fixture = TestBed.createComponent(NavigationBarGameMenuToggleComponent);
            fixture.detectChanges();
            MockGameMenuService.showGamesSubject.next(false);
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
            fixture.componentInstance.hoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
            fixture.componentInstance.stopHoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
        });

        it('toggling while hovering finalizes state - assume starting with hide menu', () => {
            const fixture = TestBed.createComponent(NavigationBarGameMenuToggleComponent);
            fixture.detectChanges();
            MockGameMenuService.showGamesSubject.next(false);
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
            fixture.componentInstance.hoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
            fixture.componentInstance.toggleGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
            fixture.componentInstance.stopHoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
        });

        it('toggling while hovering finalizes state - assume starting with showing menu', () => {
            const fixture = TestBed.createComponent(NavigationBarGameMenuToggleComponent);
            fixture.detectChanges();
            MockGameMenuService.showGamesSubject.next(true);
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
            fixture.componentInstance.hoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeTruthy();
            fixture.componentInstance.toggleGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
            fixture.componentInstance.stopHoverGameMenu();
            expect(MockGameMenuService.showGamesSubject.getValue()).toBeFalsy();
        });
    });
});
