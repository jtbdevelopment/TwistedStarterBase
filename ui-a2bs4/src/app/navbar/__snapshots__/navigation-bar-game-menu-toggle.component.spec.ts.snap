// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component:  nav bar game menu toggle component displays no toggle when player loaded 1`] = `
<app-navigation-bar-game-menu-toggle
  config={[Function MockAppConfig]}
  gameMenuService={[Function MockGameMenuService]}
  helpDisplay={[Function HelpDisplayService]}
  helpPopovers={[Function QueryList]}
  hovering="false"
  playerLoaded="false"
>
  
  
</app-navigation-bar-game-menu-toggle>
`;

exports[`Component:  nav bar game menu toggle component displays toggle when player loaded 1`] = `
<app-navigation-bar-game-menu-toggle
  config={[Function MockAppConfig]}
  gameMenuService={[Function MockGameMenuService]}
  helpDisplay={[Function HelpDisplayService]}
  helpPopovers={[Function QueryList]}
  hovering="false"
  playerLoaded={[Function Boolean]}
>
  <ul
    class="navbar-nav"
  >
    
      
    
      
    <li
      class="nav-item"
      ng-reflect-ngb-popover="Show/hide your game list for m"
      ng-reflect-placement="top"
      ng-reflect-popover-title="Games"
      ng-reflect-triggers="manual"
      ngbpopover="Show/hide your game list for more room on the screen."
      placement="top"
      popovertitle="Games"
      triggers="manual"
    >
      
          
      <a
        class="nav-link"
        href="javascript:void(0)"
      >
        
              
        <span
          class="fa fa-bars"
        />
        
          
      </a>
      
      
    </li>
    
  
  </ul>
  
</app-navigation-bar-game-menu-toggle>
`;

exports[`Component:  nav bar game menu toggle component toggling help on/off shows/hides popover 1`] = `
<app-navigation-bar-game-menu-toggle
  config={[Function MockAppConfig]}
  gameMenuService={[Function MockGameMenuService]}
  helpDisplay={[Function HelpDisplayService]}
  helpPopovers={[Function QueryList]}
  hovering="false"
  playerLoaded={[Function Boolean]}
>
  <ul
    class="navbar-nav"
  >
    
      
    
      
    <li
      aria-describedby="ngb-popover-1"
      class="nav-item"
      ng-reflect-ngb-popover="Show/hide your game list for m"
      ng-reflect-placement="top"
      ng-reflect-popover-title="Games"
      ng-reflect-triggers="manual"
      ngbpopover="Show/hide your game list for more room on the screen."
      placement="top"
      popovertitle="Games"
      triggers="manual"
    >
      
          
      <a
        class="nav-link"
        href="javascript:void(0)"
      >
        
              
        <span
          class="fa fa-bars"
        />
        
          
      </a>
      
      
    </li>
    <ngb-popover-window
      _nghost-c0=""
      class="popover bs-popover-top"
      id="ngb-popover-1"
      role="tooltip"
      style="left: 0px;"
    >
      
      
      <div
        _ngcontent-c0=""
        class="arrow"
      />
      
      
      <h3
        _ngcontent-c0=""
        class="popover-header"
      >
        Games
      </h3>
      <div
        _ngcontent-c0=""
        class="popover-body"
      >
        Show/hide your game list for more room on the screen.
      </div>
    </ngb-popover-window>
    
  
  </ul>
  
</app-navigation-bar-game-menu-toggle>
`;

exports[`Component:  nav bar game menu toggle component toggling help on/off shows/hides popover 2`] = `
<app-navigation-bar-game-menu-toggle
  config={[Function MockAppConfig]}
  gameMenuService={[Function MockGameMenuService]}
  helpDisplay={[Function HelpDisplayService]}
  helpPopovers={[Function QueryList]}
  hovering="false"
  playerLoaded={[Function Boolean]}
>
  <ul
    class="navbar-nav"
  >
    
      
    
      
    <li
      class="nav-item"
      ng-reflect-ngb-popover="Show/hide your game list for m"
      ng-reflect-placement="top"
      ng-reflect-popover-title="Games"
      ng-reflect-triggers="manual"
      ngbpopover="Show/hide your game list for more room on the screen."
      placement="top"
      popovertitle="Games"
      triggers="manual"
    >
      
          
      <a
        class="nav-link"
        href="javascript:void(0)"
      >
        
              
        <span
          class="fa fa-bars"
        />
        
          
      </a>
      
      
    </li>
    
  
  </ul>
  
</app-navigation-bar-game-menu-toggle>
`;
